types/firestore.d.ts

```diff
@@ -27,10 +27,31 @@ declare namespace FirebaseFirestore {
    */
   export type DocumentData = {[field: string]: any};
 
+  /**
+   * Similar to Typescript's `Partial<T>`, but allows nested fields to be
+   * omitted and FieldValues to be passed in as property values.
+   */
+  export type PartialWithFieldValue<T> = T extends Primitive
+    ? T
+    : T extends {}
+    ? {[K in keyof T]?: PartialWithFieldValue<T[K]> | FieldValue}
+    : Partial<T>;
+
+  /**
+   * Allows FieldValues to be passed in as a property value while maintaining
+   * type safety.
+   */
+  export type WithFieldValue<T> = T extends Primitive
+    ? T
+    : T extends {}
+    ? {[K in keyof T]: WithFieldValue<T[K]> | FieldValue}
+    : Partial<T>;
+
   /**
    * Update data (for use with [update]{@link DocumentReference#update})
    * that contains paths mapped to values. Fields that contain dots reference
-   * nested fields within the document.
+   * nested fields within the document. FieldValues can be passed in
+   * as property values.
```

Comment:
Thinking out loud: Is "property" the right term and does it make it harder to understand? Should we drop it and just call it "values"?

Reply:
I wanted to specify "property values" to convey that FieldValues can be set as the value of an object property, but not as the entire object value.